File: Data Structures & Algorithms/reorder-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        if(head == nullptr or head->next == nullptr) return;

        ListNode* slow = head;
        ListNode* fast = head;

        while(fast and fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* prev = nullptr;
        ListNode* curr = slow->next;
        slow->next = nullptr;

        while(curr){
            ListNode* tmp = curr->next;
            curr->next = prev;
            prev = curr;
            curr =tmp;
        }

        ListNode* first = head;
        ListNode* second = prev;

        while(second){
            ListNode* tmp1 = first->next;
            ListNode* tmp2 = second->next;

            first->next = second;
            second->next = tmp1;

            first = tmp1;
            second = tmp2;
        }
        return;
    }
};

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        p1 = list1 
        p2 = list2

        dummy = ListNode()    
        curr = dummy

        while p1 and p2:
            if p1.val == p2.val or p1.val < p2.val:
                curr.next = p1
                curr = p1 
                p1 = p1.next
            elif p2.val < p1.val:
                curr.next = p2
                curr = p2
                p2 = p2.next 


        if not p1:
            curr.next = p2

        if not p2:
            curr.next = p1    
        return dummy.next             
